src/psyclone/psyad/main.py

```diff
@@ -0,0 +1,143 @@
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Authors R. W. Ford and A. R. Porter, STFC Daresbury Lab
+
+'''Top-level driver functions for PSyAD : the PSyclone Adjoint
+support. Transforms an LFRic tangent linear kernel to its adjoint.
+
+'''
+from __future__ import absolute_import
+import argparse
+import logging
+from io import StringIO
+import sys
+
+import six
+
+from psyclone.psyir.frontend.fortran import FortranReader
+from psyclone.psyir.backend.fortran import FortranWriter
+from psyclone.generator import write_unicode_file
+from psyclone.psyad import generate
+
+
+def main(args):
+    '''Takes an LFRic tangent linear kernel source file as input and
+    produces its adjoint.
+
+    :param list args: the list of command-line arguments that PSyAD has \
+                      been invoked with.
+
+    '''
+    parser = argparse.ArgumentParser(
+        description="Run the PSyclone adjoint code generator on an LFRic "
+        "tangent-linear kernel file")
+    parser.add_argument(
+        '-v', '--verbose', help='increase the verbosity of the output',
+        action='store_true')
+    parser.add_argument('-oad', help='filename for the transformed code')
+    parser.add_argument('filename', help='LFRic tangent-linear source code')
+
+    args = parser.parse_args(args)
+
+    if args.verbose:
+        logging.basicConfig(level=logging.DEBUG)
+
+    # TL Fortran code
+    filename = args.filename
+    logging.info("Reading file %s", filename)
+    with open(filename) as my_file:
+        tl_fortran_str = my_file.read()
+        tl_fortran_str = six.text_type(tl_fortran_str)
+
+    ad_fortran_str = main_str(tl_fortran_str)
+
+    # AD Fortran code
+    if args.oad:
+        logging.info("Writing file %s", args.oad)
+        write_unicode_file(ad_fortran_str, args.oad)
+    else:
+        print(ad_fortran_str)
+
+
+def main_str(tl_fortran_str):
+    '''Takes an LFRic tangent-linear kernel encoded as a string as input
+    and returns its adjoint encoded as a string.
+
+    :param str tl_fortran_str: a string containing the LFRic \
```

Comment:
s/a string containing the/Fortran implementation of an/

Reply:
modified.